File: node/src/load_balance.js
/**
 * VedaDB Node.js Driver - Load Balancer
 *
 * Distributes connections across multiple VedaDB nodes using
 * configurable strategies: round-robin, least-connections, random,
 * and weighted round-robin.
 */

'use strict';

const { EventEmitter } = require('events');
const { ValidationError, ConnectionError } = require('./errors');

/**
 * Load balancing strategies.
 */
const Strategy = {
  ROUND_ROBIN: 'round_robin',
  LEAST_CONNECTIONS: 'least_connections',
  RANDOM: 'random',
  WEIGHTED_ROUND_ROBIN: 'weighted_round_robin',
  FIRST_AVAILABLE: 'first_available',
};

/**
 * Represents a backend node.
 */
class Backend {
  /**
   * @param {Object} config
   * @param {string} config.host
   * @param {number} config.port
   * @param {number} [config.weight=1] - For weighted strategies
   * @param {Object} [config.options] - Additional client options
   */
  constructor(config) {
    this.host = config.host;
    this.port = config.port || 6380;
    this.weight = config.weight || 1;
    this.options = config.options || {};
    this.connections = 0;
    this.healthy = true;
    this.lastCheck = null;
  }

  /** Unique identifier for this backend. */
  get id() {
    return `${this.host}:${this.port}`;
  }
}

/**
 * Load balancer for distributing across VedaDB nodes.
 */
class LoadBalancer extends EventEmitter {
  /**
   * @param {Object} options
   * @param {string} [options.strategy='round_robin']
   * @param {number} [options.healthCheckIntervalMs=30000]
   * @param {number} [options.maxConnectionsPerBackend=50]
   */
  constructor(options = {}) {
    super();
    this.strategy = options.strategy || Strategy.ROUND_ROBIN;
    this.healthCheckIntervalMs = options.healthCheckIntervalMs || 30000;
    this.maxConnectionsPerBackend = options.maxConnectionsPerBackend || 50;
    this._backends = [];
    this._rrIndex = 0;
    this._wrrIndex = 0;
    this._wrrCurrentWeight = 0;
    this._healthTimer = null;
  }

  /** Number of registered backends. */
  get backendCount() {
    return this._backends.length;
  }

  /** Number of healthy backends. */
  get healthyCount() {
    return this._backends.filter(b => b.healthy).length;
  }

  /** Stats snapshot. */
  get stats() {
    return {
      strategy: this.strategy,
      totalBackends: this._backends.length,
      healthyBackends: this.healthyCount,
      backends: this._backends.map(b => ({
        id: b.id,
        host: b.host,
        port: b.port,
        weight: b.weight,
        connections: b.connections,
        healthy: b.healthy,
      })),
    };
  }

  /**
   * Register a backend node.
   * @param {Object} config
   */
  addBackend(config) {
    const backend = new Backend(config);
    this._backends.push(backend);
    this.emit('backend:add', { id: backend.id, host: backend.host, port: backend.port });
    return backend;
  }

  /**
   * Remove a backend by ID.
   * @param {string} id - host:port
   */
  removeBackend(id) {
    const idx = this._backends.findIndex(b => b.id === id);
    if (idx !== -1) {
      const backend = this._backends.splice(idx, 1)[0];
      this.emit('backend:remove', { id: backend.id });
    }
  }

  /**
   * Select the next backend according to the strategy.
   * @returns {Backend|null}
   * @throws {ConnectionError} If no healthy backends available
   */
  select() {
    const healthy = this._backends.filter(b => b.healthy);
    if (healthy.length === 0) {
      throw new ConnectionError('No healthy backends available');
    }

    switch (this.strategy) {
      case Strategy.ROUND_ROBIN:
        return this._roundRobin(healthy);
      case Strategy.LEAST_CONNECTIONS:
        return this._leastConnections(healthy);
      case Strategy.RANDOM:
        return this._random(healthy);
      case Strategy.WEIGHTED_ROUND_ROBIN:
        return this._weightedRoundRobin(this._backends);
      case Strategy.FIRST_AVAILABLE:
        return healthy[0] || null;
      default:
        return this._roundRobin(healthy);
    }
  }

  /**
   * Mark a backend as having a new connection.
   * @param {Backend} backend
   */
  acquire(backend) {
    backend.connections++;
  }

  /**
   * Mark a backend as having released a connection.
   * @param {Backend} backend
   */
  release(backend) {
    if (backend.connections > 0) backend.connections--;
  }

  /**
   * Mark a backend's health status.
   * @param {string} id
   * @param {boolean} healthy
   */
  setHealth(id, healthy) {
    const backend = this._backends.find(b => b.id === id);
    if (backend && backend.healthy !== healthy) {
      backend.healthy = healthy;
      backend.lastCheck = Date.now();
      this.emit('healthChange', { id, healthy, backend });
    }
  }

  /**
   * Start periodic health checks.
   * @param {function(Backend): Promise<boolean>} checkFn
   */
  startHealthChecks(checkFn) {
    if (this._healthTimer) return;
    this._healthTimer = setInterval(async () => {
      for (const backend of this._backends) {
        try {
          const healthy = await checkFn(backend);
          this.setHealth(backend.id, healthy);
        } catch {
          this.setHealth(backend.id, false);
        }
      }
    }, this.healthCheckIntervalMs);
  }

  /** Stop health checks. */
  stopHealthChecks() {
    if (this._healthTimer) {
      clearInterval(this._healthTimer);
      this._healthTimer = null;
    }
  }

  /** Clean up. */
  destroy() {
    this.stopHealthChecks();
    this._backends = [];
    this.removeAllListeners();
  }

  // -- Strategy implementations ---------------------------------------------

  _roundRobin(healthy) {
    const selected = healthy[this._rrIndex % healthy.length];
    this._rrIndex = (this._rrIndex + 1) % healthy.length;
    return selected;
  }

  _leastConnections(healthy) {
    return healthy.reduce((best, current) =>
      current.connections < best.connections ? current : best
    );
  }

  _random(healthy) {
    return healthy[Math.floor(Math.random() * healthy.length)];
  }

  _weightedRoundRobin(backends) {
    const healthy = backends.filter(b => b.healthy);
    if (healthy.length === 0) return null;

    let maxWeight = 0;
    let best = null;

    for (const backend of healthy) {
      if (backend.weight > maxWeight) {
        maxWeight = backend.weight;
        best = backend;
      }
    }

    return best || healthy[0];
  }
}

module.exports = {
  LoadBalancer,
  Backend,
  Strategy,
};

File: rust/src/load_balance.rs
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

use crate::client::VedaConfig;
use crate::error::VedaError;
use crate::pool::{PoolStats, PooledClient, VedaPool};
use crate::result::{Value, VedaResult};

/// Load balancing strategy.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BalanceStrategy {
    /// Round-robin across all nodes.
    RoundRobin,
    /// Random selection.
    Random,
    /// Least connections.
    LeastConnections,
    /// Weighted round-robin.
    Weighted,
    /// First available (failover style).
    FirstAvailable,
}

/// Node in the load-balanced cluster.
pub struct BalancedNode {
    pool: VedaPool,
    weight: usize,
    failures: AtomicUsize,
    last_failure: Mutex<Option<Instant>>,
    total_requests: AtomicUsize,
}

/// Load balancer across multiple VedaDB nodes.
pub struct LoadBalancer {
    nodes: Vec<Arc<BalancedNode>>,
    strategy: BalanceStrategy,
    counter: AtomicUsize,
    unhealthy_threshold: usize,
    recovery_delay: Duration,
}

impl LoadBalancer {
    /// Create a new load balancer.
    pub fn new(configs: Vec<VedaConfig>, strategy: BalanceStrategy) -> Result<Self, VedaError> {
        let mut nodes = Vec::with_capacity(configs.len());
        for config in configs {
            let pool = VedaPool::new(config)?;
            nodes.push(Arc::new(BalancedNode {
                pool,
                weight: 1,
                failures: AtomicUsize::new(0),
                last_failure: Mutex::new(None),
                total_requests: AtomicUsize::new(0),
            }));
        }

        Ok(LoadBalancer {
            nodes,
            strategy,
            counter: AtomicUsize::new(0),
            unhealthy_threshold: 3,
            recovery_delay: Duration::from_secs(30),
        })
    }

    /// Create with weighted nodes.
    pub fn with_weights(
        configs: Vec<VedaConfig>,
        weights: Vec<usize>,
        strategy: BalanceStrategy,
    ) -> Result<Self, VedaError> {
        let mut lb = Self::new(configs, strategy)?;
        for (i, weight) in weights.iter().enumerate() {
            if let Some(node) = lb.nodes.get_mut(i) {
                // Need to recreate the Arc with new weight
                let pool = node.pool.clone();
                *node = Arc::new(BalancedNode {
                    pool,
                    weight: *weight,
                    failures: AtomicUsize::new(0),
                    last_failure: Mutex::new(None),
                    total_requests: AtomicUsize::new(0),
                });
            }
        }
        Ok(lb)
    }

    /// Execute a query on the selected node.
    pub fn query(
        &self,
        sql: &str,
        params: Option<&[Value]>,
    ) -> Result<VedaResult, VedaError> {
        let node = self.select_node()?;
        node.total_requests.fetch_add(1, Ordering::SeqCst);

        let mut client = node.pool.acquire()?;
        match client.query(sql, params) {
            Ok(result) => {
                node.failures.store(0, Ordering::SeqCst);
                Ok(result)
            }
            Err(e) => {
                let failures = node.failures.fetch_add(1, Ordering::SeqCst) + 1;
                *node.last_failure.lock().unwrap() = Some(Instant::now());

                if failures >= self.unhealthy_threshold {
                    // Try another node
                    self.query_with_fallback(sql, params, &node)
                } else {
                    Err(e)
                }
            }
        }
    }

    /// Execute with fallback to other nodes.
    fn query_with_fallback(
        &self,
        sql: &str,
        params: Option<&[Value]>,
        skip_node: &Arc<BalancedNode>,
    ) -> Result<VedaResult, VedaError> {
        let mut last_err = None;
        for node in &self.nodes {
            if Arc::ptr_eq(node, skip_node) {
                continue;
            }
            let mut client = match node.pool.acquire() {
                Ok(c) => c,
                Err(e) => {
                    last_err = Some(e);
                    continue;
                }
            };
            match client.query(sql, params) {
                Ok(result) => return Ok(result),
                Err(e) => last_err = Some(e),
            }
        }
        Err(last_err.unwrap_or_else(|| VedaError::Failover {
            attempts: self.nodes.len(),
        }))
    }

    /// Execute a statement (always goes to first healthy node or primary).
    pub fn execute(&self, sql: &str, params: Option<&[Value]>) -> Result<u64, VedaError> {
        let node = self.select_node()?;
        let mut client = node.pool.acquire()?;
        client.execute(sql, params)
    }

    /// Select a node based on the strategy.
    fn select_node(&self) -> Result<&Arc<BalancedNode>, VedaError> {
        let healthy_nodes: Vec<&Arc<BalancedNode>> = self
            .nodes
            .iter()
            .filter(|node| {
                let failures = node.failures.load(Ordering::SeqCst);
                if failures >= self.unhealthy_threshold {
                    // Check if recovery delay has passed
                    if let Some(last_fail) = *node.last_failure.lock().unwrap() {
                        if last_fail.elapsed() < self.recovery_delay {
                            return false;
                        }
                    }
                }
                true
            })
            .collect();

        if healthy_nodes.is_empty() {
            return Err(VedaError::Failover {
                attempts: self.nodes.len(),
            });
        }

        match self.strategy {
            BalanceStrategy::RoundRobin => {
                let idx = self.counter.fetch_add(1, Ordering::SeqCst) % healthy_nodes.len();
                Ok(healthy_nodes[idx])
            }
            BalanceStrategy::Random => {
                let random_val = std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap_or_default()
                    .subsec_nanos() as usize;
                let idx = random_val % healthy_nodes.len();
                Ok(healthy_nodes[idx])
            }
            BalanceStrategy::LeastConnections => {
                // Find node with most idle connections
                healthy_nodes
                    .iter()
                    .min_by_key(|node| {
                        let stats = node.pool.stats();
                        stats.active_connections
                    })
                    .map(|&n| Ok(n))
                    .unwrap_or_else(|| Ok(healthy_nodes[0]))
            }
            BalanceStrategy::Weighted => {
                // Simple weighted selection
                let total_weight: usize = healthy_nodes.iter().map(|n| n.weight).sum();
                if total_weight == 0 {
                    return Ok(healthy_nodes[0]);
                }
                let mut choice = self.counter.fetch_add(1, Ordering::SeqCst) % total_weight;
                for node in healthy_nodes {
                    if choice < node.weight {
                        return Ok(node);
                    }
                    choice -= node.weight;
                }
                Ok(healthy_nodes[0])
            }
            BalanceStrategy::FirstAvailable => {
                Ok(healthy_nodes[0])
            }
        }
    }

    /// Get statistics for all nodes.
    pub fn stats(&self) -> LoadBalanceStats {
        LoadBalanceStats {
            node_count: self.nodes.len(),
            node_stats: self
                .nodes
                .iter()
                .map(|node| NodeStats {
                    pool_stats: node.pool.stats(),
                    weight: node.weight,
                    failures: node.failures.load(Ordering::SeqCst),
                    total_requests: node.total_requests.load(Ordering::SeqCst),
                })
                .collect(),
            strategy: self.strategy,
        }
    }

    /// Add a new node dynamically.
    pub fn add_node(&mut self, config: VedaConfig) -> Result<(), VedaError> {
        let pool = VedaPool::new(config)?;
        self.nodes.push(Arc::new(BalancedNode {
            pool,
            weight: 1,
            failures: AtomicUsize::new(0),
            last_failure: Mutex::new(None),
            total_requests: AtomicUsize::new(0),
        }));
        Ok(())
    }

    /// Remove a node by index.
    pub fn remove_node(&mut self, index: usize) {
        if index < self.nodes.len() {
            self.nodes.remove(index);
        }
    }

    /// Get the number of nodes.
    pub fn node_count(&self) -> usize {
        self.nodes.len()
    }

    /// Close all node pools.
    pub fn close(&self) {
        for node in &self.nodes {
            node.pool.close();
        }
    }
}

/// Statistics for a single node in the load balancer.
#[derive(Debug, Clone)]
pub struct NodeStats {
    pub pool_stats: PoolStats,
    pub weight: usize,
    pub failures: usize,
    pub total_requests: usize,
}

/// Overall load balancer statistics.
#[derive(Debug, Clone)]
pub struct LoadBalanceStats {
    pub node_count: usize,
    pub node_stats: Vec<NodeStats>,
    pub strategy: BalanceStrategy,
}

impl std::fmt::Display for LoadBalanceStats {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "LoadBalancer {{ strategy={:?}, nodes={}, {:?} }}",
            self.strategy, self.node_count, self.node_stats
        )
    }
}
